### DISCUSSION TITLE
Should Dockerfile support passing in external ip using --advertiseaddr?

### DISCUSSION THREAD

**Discussion by gtallan** [LMS-Community/slimserver-platforms#46] [General]
I just converted to a Docker install of 8.3.1 and have noticed at least one place where the browser is being passed a URL using the internal ip address of the Docker container (which of course fails). I noticed this in the "Customize Device" plugin, where I believe Slim::Plugin::JiveExtras::Settings is calling Slim::Utils::Network::serverURL() to construct the URLs for the image thumbnails, and this is ultimately using hostAddr() resulting in the internal ip address being used.

I can see that Slim/Utils/Network.pm recently (2021) added an --advertiseaddr parameter, which might have been to address this kind of issue. Would it be useful to formally support and document this as a parameter in the Dockerfile? Happy to try and create a PR if this makes sense.

**Comment by gtallan** [comment]
Just verified that modifying the start-container script to include an --advertiseaddr parameter for slimserver.pl does correct this issue - the "Customize Device" settings page no longer tries to connect to 172.17.0.x, and the image thumbnails are now shown.

I just lazily hard-coded the advertiseaddr ip address for this test but will think about the correct fix.

**Comment by michaelherger** [comment]
Maybe we could add an optional variable like `$EXTRA_ARGS` to the startup script (see https://github.com/Logitech/slimserver-platforms/blob/d8120b35f0e8f3991968ab6db1abf22da89ad18f/Docker/start-container.sh#L20). It would be empty by default, but could be set in the Docker command line?

**Comment by gtallan** [comment]
I like this idea better than having code specific to --advertiseaddr - makes much more sense.

I do wonder if this has been the wrong approach to my original problem:

JiveExtras::Settings.pm is generating absolute URIs for the thumbnail images. The above fix does make these use the correct (external) ip address rather than the internal docker one. However perhaps it makes more sense just to use a relative URL, like the rest of LMS?

ie, in JiveExtras/Settings.pm (sub _addSettings), change:
my $urlBase = Slim::Utils::Network::serverURL() . "/jive$optname";
to just:
my $urlBase = "/jive$optname";

Making this change also appears to fix things correctly, at least for wallpaper images (I haven't tested sounds)

**Comment by michaelherger** [comment]
I was wondering about the relative vs. absolute URL. If you're saying it was working with the relative path, then that should be the way to go.

**Comment by Mr-win7** [comment]
I have similar request when I'm trying to use ipv6 address with the container.
Any concern about the extra args? If not I can help to create an PR for it.

**Comment by mundschenk-at** [comment]
Is `--advertiseaddr` just for the web frontend? If so, it would be better to retire it in favor of using relative URLs everywhere. If it is really about advertising (via broadcasts), it would be better to add the option for a DNS name instead of an IP address (i.e. my LMS is accessible via a TLS reverse proxy, but for its routing to work, it needs to be accessed via a hostname (SNI).

**Reply by michaelherger** [reply]
Relative URLs don't work in all cases, unfortunately.

**Reply by mundschenk-at** [reply]
Just so that I can understand, under what circumstances do relative URLs not work for a web page? I am not talking about protocols other than HTTP(S) obviously.

**Reply by michaelherger** [reply]
The web page is not the only consumer of such URLs: the players need absolute URLs, too. IIRC there is a problem with the Sounds plugin if the `advertiseaddr` isn't defined. And in the web browsers I know about plugins which do OAUTH2 with some services. In that case you have to provide an absolute URL incl. the server address as the flow would send the browser away from an LMS page, and should then redirect back to LMS.

**Reply by mundschenk-at** [reply]
Thanks, I didn't think about OAUTH2. But I still think that for most of the use cases (except maybe the players), being able to give a DNS name/base URL would be preferable.
